### PR TITLE
Command framework rework

### PR DESCRIPTION
**Pull request by JasperLorelai**
Because `VolatileCodeHandle#saveSkinData` was only ever used in the `c saveskin` command, it was refactored in favour of `VolatileCodeHandle#getSkinData`. I realise the return value is limiting, but my only intention was to remove duplicate code, the FileWriter, as the method is only used in the command anyways.

## Command framework:
The new framework uses the [ACF library by aikar](https://github.com/aikar/commands). [This line](https://github.com/JasperLorelai/MagicSpells/blob/4.0/core/build.gradle#L16) does say Paper, but it supports Spigot fine. It is a little bit more optimised for Paper.

What I'm uncertain about is how risky it would be to use the ACF Help API [here](https://github.com/JasperLorelai/MagicSpells/blob/4.0/core/src/main/java/com/nisovin/magicspells/MagicSpells.java#L559). The warnings are on their page [here](https://github.com/aikar/commands/wiki/Unstable-API).

### Format Changes:
The <spell> parameters now support spell aliases and internal spell names. The ingame names are still supported if quoted properly, but they break the tab completer, so be careful when completing the command. The tab completer now returns a spell's internal name, ingame name, and aliases.

<> - required parameter
[] - optional parameter
- `/c reload [player]` >>> `/ms reload [player]`
- `/c reloadeffectlib` >>> `/ms reloadeffectlib`
- `/c resetcd [player]` >>> `/ms resetcd [player/*] [spell/*]`
- `/c magicspellmana` >>> `/ms mana show`
- `/c resetmana <player>` >>> `/ms mana reset [player]`
- `/c setmaxmana <player> <amount>` >>> `/ms mana setmax [player] <amount>`
- `/c modifymana <player> <amount>` >>>  `/ms mana add [player] <amount>`
- `/c setmana <player> <amount>` >>> `/ms mana set [player] <amount>`
- `/c updatemanarank <player>` >>> `/ms mana updaterank [player]`
- `/c showvariable <variable> <player>` >>> `/ms variable show <variable> [player]`
- `/c modifyvariable <variable> <player> <varMod>` >>> `/ms variable modify <variable> <player> <varMod>`
- `/c magicitem <player> <magicItem> [amount]` >>> `/ms magicitem <magicItem> [amount] [player]`
- `/c download <fileName> <url>` >>> `/ms util download <url> <fileName>`
- `/c update <fileName> <url>` >>> `/ms util update <url> <fileName>`
- `/c saveskin [player]` >>> `/ms util saveskin [player]`
- `/c profilereport` >>> `/ms profilereport`
- `/c debug` >>> `/ms debug`
- `/c magicspellxp` >>> `/ms magicxp`
- `/c <spellInGame> [spellArgs]` >>> `/ms self <spell> [-p:(power)] [spellArgs]`
- `/c forcecast <player/UUID> <spellInGame> [spellArgs]` >>> `/ms as <player/UUID> <spell> [-p:(power)] [spellArgs]`
- `/c <spellInGame> <player>` >>> `/ms on <player/UUID> <spell>`
- `/c <spellInGame> <world,x,y,z,pitch,yaw>` >>> `/ms at <spell> [world] <x> <y> <z> [pitch] [yaw]`

### Permission Changes:
All listed commands no longer require OP to be performed. They are given to OP by default, but each can be granted - [the permission list](https://github.com/JasperLorelai/MagicSpells/blob/4.0/core/src/main/java/com/nisovin/magicspells/Perm.java#L24).

### CODE REVIEW

File: core/src/main/resources/config.yml [hunk 1]
```diff
@@ -5,9 +5,6 @@ general:
     enable-stat-collection: true
     enable-volatile-features: true
     enable-new-spell-handling: false
-    str-cast-usage: |
```
Comment (TheComputerGeek2) [inline core/src/main/resources/config.yml:8]: Where is the equivalent functionality or has there already been an investigation to see if this was unused?

File: core/src/main/java/com/nisovin/magicspells/MagicCommand.java [hunk 2]
```diff
@@ -0,0 +1,720 @@
+package com.nisovin.magicspells;
```
Comment (TheComputerGeek2) [inline core/src/main/java/com/nisovin/magicspells/MagicCommand.java:1]: why was everything thrown into this file? That is the opposite of the direction we want to go

File: core/src/main/resources/config.yml (hunk 1 above)
Comment (JasperLorelai) [reply core/src/main/resources/config.yml:8]: As already mentioned, there is an automated [Help API](https://github.com/aikar/commands/wiki/Command-Help) (enabled with [risks](https://github.com/aikar/commands/wiki/Unstable-API)) which handles help based on the [Description annotation](https://github.com/JasperLorelai/MagicSpells/blob/4.0/core/src/main/java/com/nisovin/magicspells/MagicCommand.java#L191).

How this issue is resolved depends on your input in the [private dev channel](https://canary.discordapp.com/channels/335237931633606656/418982700985352192/748427019775377449). I won't resolve this until that input is given.

File: core/src/main/java/com/nisovin/magicspells/MagicCommand.java (hunk 2 above)
Comment (JasperLorelai) [reply core/src/main/java/com/nisovin/magicspells/MagicCommand.java:1]: I didn't get that impression. I was following up how other large systems are handled. How do you propose I extract this?

**Comment by JasperLorelai** [comment]
I believe most, if not all previous features of subcommands have been preserved (except one - commit incoming). I have went through all subcommands and in this stage require input from other testers. The only real change is in the behaviour of the commands and their syntax - some features have been extended, some simplified (player parameter is mostly optional in most commands). In summary, I can't confirm all previous features still work on my own.

The old cast command `c` (now `ms cast self`) should be granted by default to all players, and I will fix that bug. I will make it return only the completion of spells the caster has, but I think the rest of the commands should return all spells loaded in the plugin.

I have tested the cast commands through command blocks and they work as intended. I will test the rest of the commands which logically should work. I can test the behaviour through external commands too, but I expect it to behave as intended - they're being executed like normal anyway.

As for the notice, I will make a forum thread documenting the changes and a wiki page to document the commands (will be hidden until beta 7 release). The actual notice should be given by whoever distributes the new jar.

#### TODO before possible merge:
- [ ] Permission to execute `ms cast self` should be granted by default.
- [ ] Command completion for `ms cast self` should be restricted to the owned spells in caster's spellbook.
- [ ] Test all commands in command blocks.
- [ ] Test command behaviour in External Command spell.
- [ ] Document changes (copy from PR message) to a forum thread to help construct a notice.
- [ ] Document new command list to assist testers.

**Comment by JasperLorelai** [comment]
As far as the `forcecast` alias is concerned, I made a functioning thing, but it does not work with the tab completer properly. If the syntax is invalid, it will display help for the `ms cast self` command instead. It does work if both the player and the spell arguments are given though. I think this is the best I could do. I could instead make the alias `/forcecast`.

**Comment by JasperLorelai** [comment]
A forum thread has been made with a hopefully more readable message documenting the changes made within this PR.
Link: http://magicspells.site.nfoservers.com/forums/discussion/89/

**Pull request merged**